toml/decoder.py

```diff
@@ -326,9 +349,50 @@ def loads(s, _dict=dict, decoder=None):
     multilinestr = ""
     multibackslash = False
     pos = 0
+    comment_norm_factor = 0
     for idx, line in enumerate(s):
         if idx > 0:
             pos += len(s[idx - 1]) + 1
+
+        if len(comment_info) > 0 and comment_info[0][0] - 1 == idx:
```

Comment:
I think that most of this new logic belongs in your preserve comment decoder.

Most likely, it is better to pass idx to the decoder and let the decoder determine if any of its `comment_info` have a match.

Maybe rather than `comment_info.append((a, b)); comment_lines.append(c)`, `comment_info[a].append((b, c)) if a in comment_info else comment_info[a] = [(b, c)]`

`a` needs to already be adjusted by the normalization factor for this to work though... the decoder could certainly keep track of that, I think?